File: app/src/main/java/ru/voronezhtsev/weatherapp/presentation/MainPresenter.java
package ru.voronezhtsev.weatherapp.presentation;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ru.voronezhtsev.weatherapp.data.repositories.ForecastsRepository;
import ru.voronezhtsev.weatherapp.domain.WeatherInteractor;

@InjectViewState
public class MainPresenter extends MvpPresenter<MainView> {

    private WeatherInteractor mWeatherInteractor;
    private ForecastsRepository mForecastsRepository;
    private Disposable mWeatherDisposable;
    private Disposable mForecastDisposable;

    private static final String TAG = "MainPresenter";

    MainPresenter(WeatherInteractor weatherInteractor, ForecastsRepository forecastsRepository) {
        mWeatherInteractor = weatherInteractor;
        mForecastsRepository = forecastsRepository;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        mWeatherDisposable = mWeatherInteractor.getWeather()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> {
                    getViewState().showWeather(response);
                }, throwable -> {
                    Log.d(TAG, "Error while getting current weatherInfo", throwable);
                });
        mForecastDisposable = mForecastsRepository.getForecast()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(forecastsResponse -> {
                    if (forecastsResponse.getForecast().size() >= 7) {
                        getViewState().showForecast(forecastsResponse.getForecast().subList(0, 7));
                    } else {
                        getViewState().showForecast(forecastsResponse.getForecast());
                    }
                }, error -> {
                });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        //todo Как работать c Disposable
        mWeatherDisposable.dispose();
        mForecastDisposable.dispose();
    }
}

File: app/src/main/java/ru/voronezhtsev/weatherapp/presentation/MainPresenter.kt
package ru.voronezhtsev.weatherapp.presentation

import android.annotation.SuppressLint
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ru.voronezhtsev.weatherapp.domain.ICityRepository
import ru.voronezhtsev.weatherapp.domain.IWeatherRepository
import ru.voronezhtsev.weatherapp.domain.WeatherInteractor
import ru.voronezhtsev.weatherapp.models.presentation.CityModel

@InjectViewState
class MainPresenter(private val cityRepository: ICityRepository,
                    private val weatherRepository: IWeatherRepository,
                    private val mainScreenConverter: MainScreenConverter,
                    private val weatherInteractor: WeatherInteractor,
                    private val city: Long?) : MvpPresenter<MainView>() {
    companion object {
        private const val TAG = "MainPresenter"
    }

    private val isCityModelLoaded = false

    @SuppressLint("CheckResult")
    override fun onFirstViewAttach() {

        //todo Прогрессбар пока погода грузится
        cityRepository.list.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response ->
                    val cityList = mutableListOf<CityModel>()
                    response.forEach { cityList.add(CityModel(it.id, it.name)) }
                    viewState.showNoPlacesAdded(cityList)
                    weatherInteractor.weatherList
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe({
                                viewState.showWeather(mainScreenConverter.convert(it))
                            }, {})
                }, {})
    }

    @SuppressLint("CheckResult")
    fun load(city: Long) {
        viewState.showProgressBar()
        weatherInteractor.getWeather(city)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doFinally { viewState.hideProgressBar() }
                .subscribe({
                    viewState.showWeather(mainScreenConverter.convert(it))
                }, {
                    val a = 1
                })
    }

    fun addPlace() {
        viewState.addPlace()
    }
}